core/src/main/scala/reftree/diagram/Animation.scala

```diff
@@ -119,6 +107,35 @@ object Animation {
     def iterate(n: Int)(iteration: A ⇒ A) =
       Builder((1 to n).foldLeft(frames)((current, _) ⇒ current :+ iteration(current.last)))
 
+    /** Add more frames by applying the provided iteration function until `predicate` yields true */
+    def iterateUntilAtMost(n: Int)(iteration: (A => A))(predicate: (A => Boolean)) =
+      if (frames.isEmpty) this
+      else
+        Builder(
+          (frames ++ Stream
+            .iterate(frames.last)(iteration)
+            .takeWhileInclusive(x => !predicate(x))
+            .toVector).take(n))
+
+    /** Add more frames by applying the provided iteration function until `predicate` yields true */
+    def iterateUntil(iteration: (A => A))(predicate: (A => Boolean)) =
+      if (frames.isEmpty) this
```

Comment:
For now, I'll assume reasonable usage, but I would encourage you to make some changes to prevent empty animations.